ft-add-department-teams: add departmental teams

gitau-kinuthia opened the pull request:
#### What does this PR do?
Add functionality for adding departmental teams and assigning assets to them

#### Description of Task to be completed?
- add teams model
- allow users to be  assigned to teams
- add functionality to allow assignments of assets to teams
- add functionality to view assets assigned to a team

#### How should this be manually tested?
1. On browser, goto url `http://127.0.0.1:8000/api/v1/docs/live/` to access all endpoints
2. Scroll to the `department-teams` section to view all associated endpoints for CRUD
`http://127.0.0.1:8000/api/v1/department-teams/`
`http://127.0.0.1:8000/api/v1/department-teams/<id>`

#### Any background context you want to provide?


#### What are the relevant pivotal tracker stories?
[#167427843](https://www.pivotaltracker.com/story/show/167427843)


#### Screenshots (If applicable)
![image](https://user-images.githubusercontent.com/20369585/61888472-3549ad00-af0c-11e9-9cf8-fcedccb771cb.png)

![image](https://user-images.githubusercontent.com/20369585/61888398-1d722900-af0c-11e9-998d-c85fb73bf687.png)

![image](https://user-images.githubusercontent.com/20369585/62134494-42d6ac80-b2e9-11e9-9f78-5e220817927b.png)

gitau-kinuthia committed:
ft-add-department-teams: add departmental teams

- add CRUD endpoints for teams
- allow teams to be allocated assets
- fix formatting
- add migrations to test file

[Finishes ##167427843]

gitau-kinuthia committed:
ft-add-department-teams: add departmental teams

- add CRUD endpoints for teams
- allow teams to be allocated assets
- fix formatting
- add migrations to circleci config
- increase test coverage

[Finishes ##167427843]

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,101 @@
+        )
+        data = response.json()
+        assert isinstance(data['results'], list)
+        assert response.status_code == status.HTTP_200_OK
```
It's ok to let at least one of your test method show all the dictionary key and value expected from hitting the `departmental_teams_url` endpoin in the assertions.. Then other ones could show only specific assertions particular to the test method.

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,101 @@
+        assert response.status_code == status.HTTP_200_OK
+
+    @patch("api.authentication.auth.verify_id_token")
+    def test_authenticated_user_get_specific_department_teams_no_data(
```
What does `no_data` in the test method name mean?

gitau-kinuthia replied:
The test has been renamed to `test_authenticated_user_get_specific_department_teams_with_no_assets_assigned` to be more communicative

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,101 @@
+        )
+        data = response.json()
+        assert response.status_code == status.HTTP_200_OK
+        assert "assets_assigned" in data
```
Your assertions could show the expected fields when querying for a specific department team just like it should when querying for all teams of a department.

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,101 @@
+        team_response = client.get(
+            url, HTTP_AUTHORIZATION="Token {}".format(self.token_user)
+        )
+        asset = AssetSerializer(new_asset).data
```
This is unnecessary I suppose; We could do without it.

gitau-kinuthia replied:
This test case is necessary to show the team_name field exists in the assets model. I have refactored this to a separate test to ensure test autonomy

kmazi replied:
Ok. I was referring to calling the serializer here. We could just read off the new_asset data directly for assertion.

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,101 @@
+            HTTP_AUTHORIZATION="Token {}".format(self.token_user),
+        )
+        data = response.json()
+        assert isinstance(data, dict)
```
What keys are returned after a post request? You test could show that.

kmazi commented on settings/base.py:
```diff
@@ -33,16 +33,15 @@
-DATABASES = {"default": dj_database_url.config()}
-
+# DATABASES = {"default": dj_database_url.config()}
+DATABASES = {"default": dj_database_url.config(default=config('DATABASE_URL'))}
```
Why is this necessary?

kmazi commented on core/models/centre.py:
```diff
@@ -167,3 +167,15 @@ class Meta:
         return self.name
+
+
+class DepartmentalTeam(models.Model):
```
There seem to be no tests for the expected behavior of the `DepartmentalTeam` model.

gitau-kinuthia replied:
This have been added

kmazi commented on api/tests/__init__.py:
```diff
@@ -240,6 +249,7 @@ def setUpClass(cls):
         cls.centre_url = reverse("andela-centres-list")
         cls.country_url = reverse("countries-list")
         cls.department_url = reverse("departments-list")
+        cls.departmental_teams_url = reverse("departmental-teams-list")
```
This is used only in the DepartmentTeamTestCase. We could move it there. Cleaning up this setup class can help improve the run time of tests significantly.

kmazi commented on api/tests/__init__.py:
```diff
@@ -171,6 +177,9 @@ def setUpClass(cls):
         cls.asset_assignee = apps.get_model("core", "AssetAssignee").objects.get(
             user=cls.user
         )
+        cls.asset_assignee_team = apps.get_model(
```
This can move also to the DepartmentTeamTestCase `setUp` method.

kmazi commented on api/tests/__init__.py:
```diff
@@ -59,6 +59,12 @@ def setUpClass(cls):
         cls.department_travel = apps.get_model("core", "Department").objects.create(
             name="Travel"
         )
+        cls.departmental_team = apps.get_model(
```
Both can move to the DepartmentTeamTestCase setUp method.

kmazi commented on api/serializers/andela_centres.py:
```diff
@@ -196,3 +199,56 @@ class CountrySerializer(serializers.ModelSerializer):
+
+        team_assignee = models.AssetAssignee.objects.filter(team=obj).first()
+        # The following condition introduced to make sure that results are not returned if there are orphan
+        # records in the that have no assignee (assigned_to field is optional so its possible to have an empty
```
Good doc though with slight typo error.

kmazi commented on api/serializers/andela_centres.py:
```diff
@@ -196,3 +199,56 @@ class CountrySerializer(serializers.ModelSerializer):
+            )
+            return paginated_assets.data
+        # return a standardised response to the api with the same structure as when there are results
+        return {"count": 0, "next": None, "previous": None, "results": []}
```
Nice work.

kmazi commented on .circleci/config.yml:
```diff
@@ -98,6 +98,11 @@ jobs:
+      - run:
+          name: run migrations
+          command: |
+            pipenv run python manage.py makemigrations
```
Why make migrations on circle ci again? All you need to do is run migration that has been made locally.

gitau-kinuthia replied:
Absence of this was the error that was causing my tests to fail on Circle CI. Once I added this line of code my tests could pass without throwing psycopg2 errors.

Reference : https://stackoverflow.com/questions/48184300/when-in-run-test-cases-then-i-will-get-this-error-psycopg2-operationalerror-cu?rq=1

kmazi reviewed (changes requested):
You could also show in your test that when querying the user endpoint, the response also contains the team the user belongs to.

gitau-kinuthia committed:
ft-add-department-teams: add departmental teams

- add CRUD endpoints for teams
- allow teams to be allocated assets
- fix formatting
- add migrations to circleci config
- increase test coverage
- refactor tests and apply review feedback
- remove makemigrations from circleci config file
- add migration files to git

[Finishes ##167427843]

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,171 @@
+        # results field assertions
+        assert 'name' in data['results'][0]
+        assert 'description' in data['results'][0]
+        assert 'department' in data['results'][0]
```
A better assertion would be `self.assertEqual(data["results"][0]["department"], self.department_team.department)`. Let's assert for the values of those keys rather than the keys. We can do this for the other tests.

gitau-kinuthia replied:
This has been refactored however to the unittest assertion  style

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,171 @@
+        )
+        data = response.json()
+        # return types assertions
+        assert isinstance(data['results'], list)
```
This is pytest assertion style. Would this run if the test is run with the `python manage.py test` command?

kmazi replied:
Let's make the changes so that both `pytest` command and `python manage.py test` command works for all tests.

gitau-kinuthia replied:
> This is pytest assertion style. Would this run if the test is run with the `python manage.py test` command?

Yes

gitau-kinuthia replied:
> Let's make the changes so that both `pytest` command and `python manage.py test` command works for all tests.

This has been refactored however to the unittest assertion  style

kmazi commented on api/tests/test_departmental_teams_api.py:
```diff
@@ -0,0 +1,171 @@
+        data = team_response.json()
+        assert "members" in data
+        assert len(data["members"]["results"]) == 1
+        assert data["members"]["results"][0]["email"] == new_user.email
```
Great!! These assertions are super cool.

kmazi commented on core/tests/test_department_team.py:
```diff
@@ -0,0 +1,40 @@
+    def test_cannot_add_existing_team(self):
+        team = DepartmentalTeam(name="Accounts", department_id=self.new_department.id)
+        team.save()
+        with transaction.atomic():
```
The default TestCase wraps all tests methods in a transaction. There is no real need to explicitly manage the transactions here in the tests.

gitau-kinuthia replied:
@kmazi the atomic block is actually necessary. In its absence, the following error is obtained:
```django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.```

kmazi replied:
My point is that we don't need a db transaction for this test method especially because the TestCase base class handles transaction for all test methods. Therefore calling it explicitly is unnecessary. @fgkinus

gitau-kinuthia replied:
Noted. I have been able to remove the delete function calls that were eliciting the transaction errors.

kmazi commented on core/tests/test_department_team.py:
```diff
@@ -0,0 +1,40 @@
+        team = DepartmentalTeam(name="Accounts", department_id=self.new_department.id)
+        team.save()
+        with transaction.atomic():
+            with pytest.raises(utils.IntegrityError):
```
This also is germane to pytest. Would this work with `python manage.py test`? Unless we're planning to deprecate running with django test.

gitau-kinuthia replied:
Yes the tests still run using the ```python manage.py test```

gitau-kinuthia replied:
This has been refactored however to the unittest assertion  style

gitau-kinuthia committed:
ft-add-department-teams: add departmental teams

- Fix test assertions to be more specific
- Use unittest assertion style
- remove atomic transactions

[Delivers ##167427843]
